packages/jamtools/features/modules/dashboards/keytar_and_foot_dashboard/multi_octave_supervisor.tsx

```diff
@@ -0,0 +1,161 @@
+import React from 'react';
+
+import {ModuleAPI} from '~/core/engine/module_api';
+import {MidiEvent, MidiEventFull} from '~/core/modules/macro_module/macro_module_types';
+import {Button} from '~/core/components/Button';
+
+type MultiOctaveSupervisorMidiState = {
+    currentlyHeldDownInputNotes: MidiEvent[];
+};
+
+export class MultiOctaveSupervisor {
+    private macros!: Awaited<ReturnType<MultiOctaveSupervisor['createMacros']>>;
+    private states!: Awaited<ReturnType<MultiOctaveSupervisor['createStates']>>;
+    private actions!: ReturnType<MultiOctaveSupervisor['createActions']>;
+
+    private midiState: MultiOctaveSupervisorMidiState = {
+        currentlyHeldDownInputNotes: [],
+    };
+
+    private handleKeyboardNote = async (fullEvent: MidiEventFull) => {
+        const event = fullEvent.event;
+
+        this.macros.pagedOctaveOutput.send(event);
+
+        if (event.type === 'noteon') {
+            if (!this.midiState.currentlyHeldDownInputNotes.find(e => e.number === event.number)) {
+                this.midiState = {
+                    ...this.midiState,
+                    currentlyHeldDownInputNotes: [
+                        ...this.midiState.currentlyHeldDownInputNotes,
+                        event,
+                    ],
+                };
```

Comment:
We should normally use immer to avoid the spread operator, but this is critical path local midi state, so we don't want to do any function calls if unnecessary. It's not very complicated/large state, so it's manageable in that sense too.